aderyn_core/src/detect/high/pre_declared_variable_usage.rs

```diff
@@ -0,0 +1,133 @@
+use std::collections::{BTreeMap, HashSet};
+use std::convert::identity;
+use std::error::Error;
+
+use crate::ast::{ASTNode, NodeID};
+
+use crate::capture;
+use crate::context::browser::{
+    AppearsAfterNodeLocation, ExtractIdentifiers, ExtractVariableDeclarations,
+};
+use crate::detect::detector::IssueDetectorNamePool;
+use crate::{
+    context::workspace_context::WorkspaceContext,
+    detect::detector::{IssueDetector, IssueSeverity},
+};
+use eyre::Result;
+
+// HOW TO USE THIS TEMPLATE:
+// 1. Copy this file and rename it to the snake_case version of the issue you are detecting.
+// 2. Rename the PreDeclaredLocalVariableUsageDetector struct and impl to your new issue name.
+// 3. Add this file and detector struct to the mod.rs file in the same directory.
+// 4. Implement the detect function to find instances of the issue.
+
+#[derive(Default)]
+pub struct PreDeclaredLocalVariableUsageDetector {
+    // Keys are: [0] source file name, [1] line number, [2] character location of node.
+    // Do not add items manually, use `capture!` to add nodes to this BTreeMap.
+    found_instances: BTreeMap<(String, usize, String), NodeID>,
+}
+
+impl IssueDetector for PreDeclaredLocalVariableUsageDetector {
+    fn detect(&mut self, context: &WorkspaceContext) -> Result<bool, Box<dyn Error>> {
+        // Since this is restricted to local variables, we examine each function independently
+        for function in context
+            .function_definitions()
+            .into_iter()
+            .filter(|&f| f.implemented)
+        {
+            let local_variable_declaration_ids = ExtractVariableDeclarations::from(function)
+                .extracted
+                .iter()
+                .map(|vd| vd.id)
+                .collect::<HashSet<_>>();
+
+            let used_local_variables = ExtractIdentifiers::from(function).extracted;
+
+            let used_local_variables = used_local_variables
+                .iter()
+                .filter(|identifier| {
+                    identifier
+                        .referenced_declaration
+                        .is_some_and(|referenced_declaration| {
+                            local_variable_declaration_ids.contains(&referenced_declaration)
+                        })
+                })
+                .collect::<HashSet<_>>();
+
+            for used in used_local_variables {
+                if let Some(id) = used.referenced_declaration {
+                    if let Some(ASTNode::VariableDeclaration(variable_declaration)) =
+                        context.nodes.get(&id)
+                    {
+                        if !used
+                            .appears_after(context, variable_declaration)
+                            .is_some_and(identity)
+                        {
+                            capture!(self, context, used);
+                        }
```

Comment:
Apparently, context.get_source_unit_for_child(node) returns None for these 2 nodes ! So we panic fail